packages/protocol/contracts/bridge/TokenVault.sol

```diff
@@ -253,9 +248,14 @@ contract TokenVault is EssentialContract {
                 name: t.name()
             });
 
-            uint256 _balance = t.balanceOf(address(this));
-            t.safeTransferFrom(msg.sender, address(this), amount);
-            _amount = t.balanceOf(address(this)) - _balance;
+            if (token == resolve("taiko_token", true)) {
+                IMintableERC20(token).burn(msg.sender, amount);
+                _amount = amount;
+            } else {
+                uint256 _balance = t.balanceOf(address(this));
+                t.safeTransferFrom(msg.sender, address(this), amount);
+                _amount = t.balanceOf(address(this)) - _balance;
```

Comment:
For some tokens, if you transfer(x) amount, it may end up sending the recipient y amount (y <x) but deduct x amount from the sender. Therefore, we here must measure how much were received, not how much were sent?